### PR TITLE
Feat: event factory support

### PR DESCRIPTION
**Pull request by kares**
Introduce a `new_event` (factory) interface for the Ruby plugin API.

The scope of this PR is to introduce the `EventFactory` interface with a single method `EventFactory#new_event` and two implementations (perhaps `BasicEventFactory` and `TargetedEventFactory`), and to effectively add three methods to the API of `LogStash::Plugin`:

 - `Plugin#event_factory`: returns a _memoized_ `EventFactory` instance, that can be used to create events
 - `Plugin#targeted_event_factory`: returns a _memoized_ `EventFactory` that respects the plugin's `target` configuration.
   - if no `target` is provided, this effectively is equivalent to `Plugin#event_factory`.
   - if a `target` is provided, returns a `TargetedEventFactory` that wraps the return value of `Plugin#event_factory`.
 - `Plugin#create_event_factory`: returns a _new_ `EventFactory` based on the plugin's configuration and state. We may want to mark this as an "internal" or "advanced" API. 

#### NOTES (from @yaauie)

 - We should document these methods as effectively-final, warning plugin developers from overriding them. There are a _lot_ of possible sharp edges here, and any overriding will create a tight coupling that eventually prevents us from moving forward.
 - The `Plugin#create_event_factory` can probably be a _private_ part of the API. I see no reason why a _plugin_ would need to provide its own implementation, and doing so would prevent us from changing the internals in logstash core to support configuration-driven alternate event factories (such as factories that produce copy-on-write events, alternate field-reference-syntax events, etc.).
 - we should memoize to underscore-prefixed ivars in a concurrency-safe manner and document the ivar as an implementation detail that should not be relied on (such as `@_event_factory` and `@_targeted_event_factory`). 

## Release notes

[rn:skip]

## What does this PR do?

This PR is motivated by having a way of re-using `LogStash::Event.from_json` with `target => [namespace]` (e.g. in the JSON codec). We [benchmarked](https://gist.github.com/kares/03fee52a0618e4b94a8dc4fcdd52b68d#file-_results-out) the generic `LogStash::Json` parsing (which has JRuby parsing hooks) and found it noticeably slower compared to the Java native `Event.fromJson`.

A clean way forward would be to introduce a notion of an `EventFactory` with a `newEvent` (`new_event`) interface.
The factory is meant to be later re-used/extended with ECS compatibility (`new_event` behaving slightly differently in legacy vs ecs mode), this is out-of-scope here, still something to keep in mind.

## Why is it important/What is the impact to the user?

It impacts plugin development which gets simpler as we'll provide a recommended way to create new events.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] ~~I have made corresponding changes to the documentation~~
- [x] ~~I have made corresponding change to the default configuration files (and/or docker env variables)~~
- [x] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [x] few micro benchmarks have been run, see **performance numbers at https://github.com/logstash-plugins/logstash-mixin-event_support/pull/2**

## Use cases

Sample use-case (using the [adapter + from_json helper](https://github.com/logstash-plugins/logstash-mixin-event_support/pull/2/files#diff-e92f351c26e57f5aa73e867d73f68d9929559bb235c90a67a930028eecea4fa4)) in the [json codec](https://github.com/logstash-plugins/logstash-codec-json/pull/37/files#diff-342311fbc3c5f7af21e6a0511156a1fc606c37a4d9c998c11d52b4daa7a418efR69).

meta-issue: https://github.com/elastic/logstash/issues/13021

### CODE REVIEW

File: logstash-core/src/main/java/org/logstash/EventFactory.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+package org.logstash;
+
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * A factory for events.
+ */
+@FunctionalInterface
+public interface EventFactory {
```
Comment (yaauie) [inline logstash-core/src/main/java/org/logstash/EventFactory.java:29]: This new interface _very nearly_ is implemented by the `co.elastic.logstash.api.EventFactory` that is available to java-based plugins through `co.elastic.logstash.api.Context` which is provided to the constructor. Can we converge without breaking consumers of `co.elastic.logstash.api.EventFactory`?

**Review by yaauie — COMMENTED** [review]
A few notes from my review of your draft, in line.

Before this gets merged, I'd also like to see a draft of our `EventFactorySupport` adapter in-flight to ensure that we can expose an approximation of this functionality to updated plugins running on older Logstashes. 

A possible sharp edge I can see is that if for example our JSON codec provides a block "event factory" to an older Logstash that doesn't acknowledge the block, will the block simply be ignored? Most of our support adapters target the Plugin API (and can thus be scoped to the plugin that includes them), but this requires a change to the Event API. Perhaps we can solve this by providing version-based fallback behaviour in the JSON codec. Or by using ruby refinements on `Event`?

File: logstash-core/lib/logstash/plugins/event_factory_support.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        end
+
+      end
+      private_constant :BasicEventFactory
+
+      class TargetedEventFactory
+
+        def initialize(inner, target)
+          @delegate = inner
+          @target = target
```
Comment (kares) [inline logstash-core/lib/logstash/plugins/event_factory_support.rb:50]: TODO: needs reference validation logic

File: logstash-core/src/main/java/org/logstash/EventFactory.java (hunk 1 above)
Comment (kares) [reply logstash-core/src/main/java/org/logstash/EventFactory.java]: @yaauie :100: thanks I completely missed the existing `EventFactory` - for sure, no reason to re-invent the wheel!
the only down-side is `org.logstash.Event.fromJson` assumes the concrete implementation being returned, thus we'll assume and cast the returned events.

**Comment by kares** [comment]
> Before this gets merged, I'd also like to see a draft of our `EventFactorySupport` adapter in-flight to ensure that we can expose an approximation of this functionality to updated plugins running on older Logstashes.

here's the adapter **[falling back](https://github.com/logstash-plugins/logstash-mixin-event_support/pull/2/files#diff-e92f351c26e57f5aa73e867d73f68d9929559bb235c90a67a930028eecea4fa4R28) on LS that does not `defined? LogStash::Plugins::EventFactorySupport`**

> A possible sharp edge I can see is that if for example our JSON codec provides a block "event factory" to an older Logstash that doesn't acknowledge the block, will the block simply be ignored?

correct, the block is passed to the method implicitly for an old LS (and is never yielded to) : 
```
>> LOGSTASH_VERSION
=> "7.13.0"
>> LogStash::Event.from_json('{ "a": "b" }') { |data| raise "unexpected" }
=> [#<LogStash::Event:0x34fb5c3f>]
```

> Most of our support adapters target the Plugin API (and can thus be scoped to the plugin that includes them), but this requires a change to the Event API. Perhaps we can solve this by providing version-based fallback behaviour in the JSON codec. Or by using ruby refinements on `Event`?

the approach would be we wrap `from_json {}` block form and provide a new method to be consumed by the codecs that are going to need `Event.from_json(json) { |data| event_factory.new_event(data) }` this way we pretty much only have a single place where the block form is known (thus the idea to not make it official API yet but I know that it gets semi-official anyway :stuck_out_tongue_winking_eye:) - details are part of the event_support mixin **PR to provide an [opt-in `FromJsonHelper` module](https://github.com/logstash-plugins/logstash-mixin-event_support/pull/2/files#diff-7618f93a38070c4b724444abfad705ab46dcaafff1185fe4cd83d8413cd6a8fa)**.

File: logstash-core/src/main/java/org/logstash/plugins/ContextImpl.java [hunk 3]
```diff
@@ -68,19 +64,6 @@ public DeprecationLogger getDeprecationLogger(Plugin plugin) {
...
-
-            @Override
-            public Event newEvent(Map<? extends Serializable, Object> data) {
-                if (data instanceof ConvertedMap) {
-                    return new org.logstash.Event((ConvertedMap)data);
-                }
-                return new org.logstash.Event(data);
-            }
-        };
+        return BasicEventFactory.INSTANCE;
```
Comment (yaauie) [inline logstash-core/src/main/java/org/logstash/plugins/ContextImpl.java:67]: I can see why adding an `EventFactory getTargetedEventFactory()` for the Java API is out of scope here -- the context doesn't have the plugin's config, and we don't have standardized ways of defining plugin options.

A native plugin could also _wrap_ the `EventFactory` that they retrieve from here, so it isn't that big of a deal anyway.

But we should file a follow-up task to make sure that we aren't leaving the Java API behind.

File: logstash-core/lib/logstash/plugins/event_factory_support.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
```
Comment (yaauie) [inline logstash-core/lib/logstash/plugins/event_factory_support.rb:50]:
```suggestion
        def initialize(inner, target)
          fail(ArgumentError,"invalid EventFactory `#{inner}`") unless inner.respond_to?(:new_event)
          fail(ArgumentError,"Invalid target field reference `#{target}`") unless org.logstash.FieldReference.isValid(target)
          @delegate = inner
          @target = target
```

**Review by yaauie — COMMENTED** [review]
This is fantastic, especially along-side the [support adapter](https://github.com/logstash-plugins/logstash-mixin-event_support/pull/2#pullrequestreview-690868161).

Do we want the block form to be the _only_ way to use the event factory with ruby's `Event#from_json`?
   - Would it make sense to also support `Event#from_json(payload, event_factory)`?
   - Or can we count on the `&` to-proc magic of `Event#from_json(payload, &event_factory)` working?

I have left a few comments, but have no major issues remaining and am confident that we can get these through in short order.

**Comment by kares** [comment]
> Do we want the block form to be the only way to use the event factory with ruby's `Event#from_json`?

My thinking was let's do the block form as it's more flexible + avoids (more) confusion about the dual nature of things - Java Event vs Ruby Event, Java EventFactory vs Ruby factories.

>  - Would it make sense to also support `Event#from_json(payload, event_factory)`?

It would if the manual `RubyEvent` wrapped for the `org.logstash.Event` wasn't "invented" and we relied on JRuby's Java integration. Which is something we can attempt in the future ...

For now, since the runtime does not know `RubyEvent` is just a wrapper for the Java `Event`, we would still need to play tricks to wrap the `event_factory` to be able to pass it as a Java `EventFactory` (doing `RubyEvent` unwrapping and re-wrapping like currently in the code with the block-form). 

> - Or can we count on the & to-proc magic of Event#from_json(payload, &event_factory) working?

As explained above we unfortunately can not atm, if there was only a Java `EventFactory` and `RubyEvent = org.logstash.Event` than `Event#from_json` would auto-magically work with a block as an implementation for the functional interface. 

If we ever bridge the gap between 2 types of events than the current block form doing `{ event_factory.new_event }` would continue working as a Java `EventFactory` implementation and the next step from there would be to close the gap with dropping the Ruby factories in favor of the Java ones. This would be part of the reasoning to prefer a block instead of the second `event_factory` argument.

File: logstash-core/lib/logstash/plugins/event_factory_support.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      lazy_init_attr :targeted_event_factory, variable: :@_targeted_event_factory do
+        raise ArgumentError.new('config(:target) not present') unless respond_to?(:target)
+        target.nil? ? event_factory : TargetedEventFactory.new(event_factory, target)
+      end
+
+      private
+
+      # @api private
+      # @since LS 7.14
+      def create_event_factory
```
Comment (kares) [inline logstash-core/lib/logstash/plugins/event_factory_support.rb:30]: do we drop this method for now? to align with the [support mixin](https://github.com/logstash-plugins/logstash-mixin-event_support/pull/2#discussion_r657261492)

Comment (yaauie) [reply logstash-core/lib/logstash/plugins/event_factory_support.rb:30]: It's private, so not necessarily something we need to align on. It is also a good hook point for future efforts in LS core that don't need backport support (things like alternate `Event` implementation experiments like copy-on-write or field reference parser changes), so I expect even if we remove it now we will add it back in short order.

Comment (kares) [reply logstash-core/lib/logstash/plugins/event_factory_support.rb:30]: okay, sticking around just wanted to double check given what `private` means in Ruby land :wink:

**Review by yaauie — APPROVED** [review]
LGTM 🎉

**Pull request merged**